contracts/products/SoteriaCoverageProduct.sol

```diff
@@ -0,0 +1,681 @@
+
+    /// @notice Registry contract.
+    IRegistry internal _registry;
+
+    /// @notice Cannot buy new policies while paused. (Default is False)
+    bool internal _paused;
+
+    /***************************************
+    BOOK-KEEPING VARIABLES
+    ***************************************/
+
+    /// @notice The total policy count.
+    uint256 internal _totalPolicyCount;
+
+    /**
+     * @notice The maximum rate charged per second per wei of coverLimit.
+     * @dev Because Solidity cannot store fractions in a single variable, need two variables: one for numerator and one for divisor
+     * @dev We also need to be careful to perform multiplication before division, as Solidity rounds down to 0
+     * @dev For testing assume _maxRate reflects 10% of coverLimit annually = 1/315360000
+     */
+    uint256 internal _maxRateNum;
+    uint256 internal _maxRateDenom;
+
+    /// @notice Maximum epoch duration over which premiums are charged.
+    uint256 internal _chargeCycle;
+
+    /**
+     * @notice The cooldown period
+     * Withdrawing total soteria account balance is a two-step process
+     * i.) Call deactivatePolicy(), which sets cover limit to 0 and begins a cooldown time
+     * ii.) After the cooldown period has been completed, the policy holder can then withdraw funds
+     * @dev We could use uint40 to store time, I thought it easier to just use uint256s in this contract. We're also not packing structs in this contract.
+     */
+    uint256 internal _cooldownPeriod;
+
+    /**
+     * @notice Percentage of cover limit purchased, that will be rewarded for using a referral code
+     * @dev In units of bps, or out of 10,000 parts. Default is 5% or _referralRewardPercentage = 500.
+     */
+    uint256 internal _referralRewardPercentage;
+
+    /**
+     * @notice Policy holder address => Timestamp that a depositor's cooldown started
+     * @dev this is set to 0 to reset (default value is 0 in Solidity anyway)
+     */
+    mapping(address => uint256) internal _cooldownStart;
+
+    /// @notice The policyholder => Soteria account balance.
+    mapping(address => uint256) internal _accountBalanceOf; // Considered _soteriaAccountBalance name
+
+    /// @notice The policyholder => policyID.
+    mapping(address => uint256) internal _policyOf;
+
+    /// @notice The cover limit for each policy(policyID => coverLimit).
+    mapping (uint256 => uint256) internal _coverLimitOf;
+
+    /// @notice This mapping is created for the purpose of avoiding the unintended side effect where `user deactivates policy -> cover limit set to 0 -> minRequiredAccountBalance also set to 0 unintentionally because minRequiredAccountBalance = coverLimit * chargeCycle * maxRate -> user can withdraw all funds immediately after deactivating account
+    /// @dev This mapping is intended to mirror the _coverLimitOf mapping, except time between cooldown start and after cooldown completed
+    mapping (uint256 => uint256) internal _preDeactivateCoverLimitOf;
```

Comment:
Do we need this mapping for every policyholder?

Reply:
This was for the an edge case. This is the discord link https://discord.com/channels/832767444488945725/895140641879760906/931066114681745438

I tried to remove it but couldn't find solution for it yet.

Reply:
I think this mapping is also the solution to the edge case you pointed out here @nikitabuzov - https://github.com/solace-fi/solace-core/pull/246#discussion_r792285578

Reply:
Ok, let's keep it as is for now, but feels a bit redundant